File: crates/ffi/cpp/src/gaussian.cpp
#include "gaussian.hpp"

double gaussian_log_prob(rust::Slice<const double> x) {
  const double *ptr = x.data();
  size_t n = x.size();

  double s = 0.0;
  for (size_t i = 0; i < n; i++) {
    s -= 0.5 * ptr[i] * ptr[i];
  }
  return s;
}

void gaussian_grad(rust::Slice<const double> x, rust::Slice<double> grad) {
  const double *xp = x.data();
  double *gp = grad.data();
  size_t n = x.size();

  for (size_t i = 0; i < n; i++) {
    gp[i] = -xp[i];
  }
}

File: crates/ffi/src/gaussian.rs
use crate::ffi;
use kernels::density::{GradLogDensity, LogDensity};

#[derive(Debug, Clone, Copy, Default)]
pub struct Gaussian;

impl LogDensity for Gaussian {
    fn log_prob(&self, x: &[f64]) -> f64 {
        ffi::gaussian_log_prob(x)
    }
}

impl GradLogDensity for Gaussian {
    fn grad_log_prob(&self, x: &[f64], grad: &mut [f64]) {
        ffi::gaussian_grad(x, grad)
    }
}
